subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/TargetType.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath.intercept;
+
+import org.codehaus.groovy.reflection.CachedClass;
+import org.codehaus.groovy.reflection.ReflectionCache;
+import org.gradle.api.NonNullApi;
+
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper for {@code Class} instances that knows about Groovy type coercions.
+ *
+ * @param <T> the target type
+ */
+@NonNullApi
+public class TargetType<T> {
+    private final CachedClass cachedClass;
+
+    private TargetType(Class<T> cls) {
+        cachedClass = ReflectionCache.getCachedClass(cls);
+    }
+
+    /**
+     * Checks if the given argument is assignment-compatible with this type, taking Groovy coercions into account.
+     *
+     * @param argument the argument, can be {@code null}
+     * @return {@code true} if the argument can be assigned to a variable of this type
+     */
+    public boolean isAssignableFrom(@Nullable Object argument) {
+        return argument == null || cachedClass.isAssignableFrom(argument.getClass());
```

Comment:
Apparently, there's more to that: `String[]` is not assignable from `GString[]`, but the latter still can be coerced to it. So this implementation misses, for example, invocation of `ProcessGroovyMethods.execute(GString[])` which is allowed in Groovy (it binds to `ProcessGroovyMethods.execute(String[])`). Interestingly, `GString[].execute()` is not allowed, and receivers aren't coerced in general. GString just delegates all its missed method to String, and that's how `GString.execute` works (it binds to `ProcessGroovyMethods.execute(String)` through `String.execute()`.

I'll fix this and re-open the PR for review.